src/remote-editor-service.ts

```diff
@@ -0,0 +1,268 @@
+  kind: 'resolution';
+  resolution: any;
+}
+interface Rejection {
+  kind: 'rejection';
+  rejection: string;
+}
+
+interface Deferred<V> {
+  resolve: (resp: V) => void;
+  reject: (err: any) => void;
+  promise: Promise<V>;
+}
+
+type Request = InitRequest | FileChangedRequest | GetWarningsRequest |
+    GetDocumentationRequest | GetDefinitionRequest |
+    GetTypeaheadCompletionsRequest | ClearCachesRequest;
+interface InitRequest {
+  kind: 'init';
+  basedir: string;
+}
+interface FileChangedRequest {
+  kind: 'fileChanged';
+  localPath: string;
+  contents?: string;
+}
+interface GetWarningsRequest {
+  kind: 'getWarningsFor';
+  localPath: string;
+}
+interface GetDocumentationRequest {
+  kind: 'getDocumentationFor';
+  localPath: string;
+  position: Position;
+}
+interface GetDefinitionRequest {
+  kind: 'getDefinitionFor';
+  localPath: string;
+  position: Position;
+}
+interface GetTypeaheadCompletionsRequest {
+  kind: 'getTypeaheadCompletionsFor';
+  localPath: string;
+  position: Position;
+}
+interface ClearCachesRequest {
+  kind: 'clearCaches';
+}
+
+/**
+ * Runs this file (remote-editor-service.js) in a new node process and
+ * exposes a promise based request API for communicating with it.
+ */
+class SelfChannel {
+  private _child: child_process.ChildProcess;
+  private _idCounter = 0;
+  private _outstandingRequests = new Map<number, Deferred<any>>();
+  constructor() {
+    this._child = child_process.fork(__filename, [], {});
+    this._child.addListener(
+        'message', (m: ResponseWrapper) => this._handleResponse(m));
+  }
+
+  async request(req: Request): Promise<any> {
+    const id = this._idCounter++;
+    const deferred = makeDeferred<any>();
+    this._outstandingRequests.set(id, deferred);
+    await this._sendRequest(id, req);
+    return deferred.promise;
+  }
+
+  private _handleResponse(response: ResponseWrapper): void {
+    const deferred = this._outstandingRequests.get(response.id);
+    if (!deferred) {
+      return;
+    }
+    switch (response.value.kind) {
+      case 'resolution':
+        return deferred.resolve(response.value.resolution);
+      case 'rejection':
+        return deferred.reject(response.value.rejection);
+      default:
+        const never: never = response.value;
+        throw new Error(`Got unknown kind of response: ${util.inspect(never)}`);
+    }
+  }
+
+  private async _sendRequest(id: number, value: Request): Promise<void> {
+    const request: RequestWrapper = {id, value: value};
+    await new Promise((resolve, reject) => {
+      (<any>this._child.send)(
+          request, (err: any) => err ? reject(err) : resolve());
+    });
+  }
+
+  dispose(): void {
+    this._child.kill();
+  }
+}
+
+/**
+ * Provides a similar interface to EditorServer, but implemented out of process.
+ *
+ * This class runs in-process and communicates via SelfChannel with
+ * EditorServer, which runs in the child process.
+ */
+export class RemoteEditorService extends BaseEditor {
```

Comment:
There are several meaningful classes in this file, it might be nice to put them in separate files in a folder.